src/algebra/bit manipulation.md

```diff
@@ -0,0 +1,214 @@
+
+$4_2 << 1 = 100 << 1 = 1000$
+
+$1000 = 8_2$ and clearly, $4 * 2 = 8$
+
+So, for a division by two, it would be exactly the same, but using the operator $<<$.
+
+$4_2 >> 1 = 100 >> 1 = 10$
+
+In this case, we would remove one bit from the end, this is what is called the **least significant bit**, the rightmost bit of the representation.
+
+Of course, the **most significant bit** is the leftmost bit.
+
+And then, how could it carry out operations with other powers, greater than two?
+
+Easy, you just have to remember the exponent of two, to create said power. In other words:
+
+$4 = 2 ^ 2$, so to do multiplication or division operations with **4**, we must move **2 bits**.
+
+$8 = 2 ^ 3$, so for **8** it would be **3 bits**. *And so on*.	
+
+
+* ## Bitwise operators.
+
+ 	* **&** : The bitwise AND operator compares each bit of its first operand with the corresponding bit of its second operand. If both bits are 1, the corresponding result bit is set to 1. Otherwise, the corresponding result bit is set to 0.
+ 	
+ 	*  **|** : The bitwise inclusive OR operator compares each bit of its first operand with the corresponding bit of its second operand. If one of the two bits is 1, the corresponding result bit is set to 1. Otherwise, the corresponding result bit is set to 0.
+
+	* **^** : The bitwise exclusive OR (XOR) operator compares each bit of its first operand with the corresponding bit of its second operand. If one bit is 0 and the other bit is 1, the corresponding result bit is set to 1. Otherwise, the corresponding result bit is set to 0.
+
+Examples:
+
+```
+  11111111 11110000
+& 00000000 01100011
+  _________________
+  00000000 01100000
+```
+
+```
+  11111111 11110000
+| 00000000 01100011
+  _________________
+  11111111 11110011
+```
+
+```
+  11111111 11110000
+^ 00000000 01100011
+  _________________
+  11111111 10010011
+```
+
+# Useful tricks.
+
+## Check if a bit is on (to one)
+
+The bit's value can be obtained by shifting the number to the right x times and then bit AND ( & ) on it:
+
+``` cpp
+bool check(int number, int bit) {
+    return ((number >> bit) & 1);
+}
+```
+
+## Change the state of a bit n to x
+
+``` cpp
+number ^= (-x ^ number) & (1 << n);
+```
+
+## Even or Odd
+
+Function to know if a number is even or odd.
+
+True if number is odd, false for number is even.
+
+``` cpp
+bool check(int number) {
+    return (number & 1);
+}
+```
+
+## Check if an integer is a power of 2
+
+``` cpp
+bool powerTwo = n && !(n & (n - 1));
+```
+
+## Brian Kernighan's algorithm.
+
+The idea is to consider only the set bits of an integer by turning off its rightmost set bit (after counting it), so the next iteration of the loop considers the Next Rightmost bit.
+
+The expression **n & (n-1)** can be used to turn off the rightmost set bit of a number **n**. This works like the expression **n-1** flips all bits after the rightmost set bit of n, including the rightmost set bit. Therefore, **n & (n-1)** returns the last flipped bit of **n**.
```

Comment:
An example would be nice here. Not that easy to understand via just text.
But an example is obvious.

```
n         = 01011000
n-1       = 01010111
--------------------
n & (n-1) = 01010000
```